### DISCUSSION TITLE
Unity UPM support

### DISCUSSION THREAD

**Discussion by loofou** [exodrifter/unity-rumor#73] [General]
Hey, 

not sure you are still actively supporting this, but in case you do: Would you be able to add support for the Unity URP package management? Using git submodules is really not the way to go with Unity and the old unity assets are also nearly deprecated. So for future-proofing it makes sense to support the latest package system for Unity.

From what I understand you simply have to add a package meta file to your repo with some information about the package and it should "just work".

Otherwise your system looks great! I was looking for something like it for a while, so I'm going to give i it a try in my project.

Thanks,
Loofou

**Comment by exodrifter** [comment]
When I first started working on Rumor and wrote the documentation for the first time, URP did not exist yet and that's why the documentation recommends submodules. I agree with you -- submodules don't seem like the best approach now. I've [experimented with URP](https://github.com/exodrifter/unity-aural/blob/master/package.json) some time ago and I believe you're right about needing just the `package.json` and `package.json.meta` files in the repository for it to work. Although, I remember being unhappy with how awkward it was to add a repository to your Unity project... maybe that's fixed by now?

I'm happy you are looking forward to using Rumor! I still want to support and maintain it but I am not actively doing so right now since I've decided to move to using Godot for game development and most of my free time has been consumed with learning a new tool and ecosystem. More than happy to review a pull request though, if you feel so inclined. Otherwise, I can probably find a free hour or two to do it in a few weeks.

As a warning, there are several things about the syntax I am unhappy with (such as the inconsistent syntax for conditionals, string substitutions, assignment, etc) so things might change a tiny bit when I get back around to active development. There's also other things I want to improve, like the compilation messages being pretty useless. And of course, there are many features I wish to add such as localization support. If you find my scripting language lacking (as it certainly is), I recommend the much more thoroughly developed [YarnSpinner](https://github.com/YarnSpinnerTool/YarnSpinner) which has a similar design philosophy.

If you do finish making a game with Rumor, let me know and I'll add it to [the list](https://github.com/exodrifter/unity-rumor/wiki/Games-made-with-Rumor). :heart:

**Reply by loofou** [reply]
Hey, thanks for the answer!

I can certainly create a pull request with the urp changes (I also found some minor issues in the wiki).

About the scripting syntax and expression layer: I actually want to use rumor embedded in my own lua-based scripting layer. Yarn and Ink both do not support runtime compilation, so that's why I was looking at rumor.

So a limited scripting system in rumor might not actually be that big of a problem right now. But well, I'm still experimenting.

My end goal is to procedurally generate storylets and having them handled in a smart way. The decision and generation logic is here, i just need something for the actual runtime dialog handling, as I'm sure there is something out there that will safe me time in writing my own.

**Reply by exodrifter** [reply]
Ah, I see. Runtime compilation is a feature of Rumor I'm quite happy with, but I haven't spent a lot of time optimizing the compilation (and I will likely have little interest in doing so in the near future), so you may need to do the compilation ahead of time and cache the results. Though, for most situations I've encountered, Rumor is plenty fast as it is already.

Using procedural generation for the dialog is actually one of the things I wanted to experiment with in Rumor, so I'm interested in hearing how it goes! Do feel free to report back with your findings :mag_right:

**Comment by loofou** [comment]
PR is ready: https://github.com/exodrifter/unity-rumor/pull/74
